esp-hal/src/gpio/mod.rs

```diff
@@ -2250,60 +2404,117 @@ mod asynch {
             self.pin.wait_for_rising_edge().await
         }
 
         /// Wait for the pin to undergo a transition from high to low.
+        ///
+        /// See [Self::wait_for] for more information.
         pub async fn wait_for_falling_edge(&mut self) {
             self.pin.wait_for_falling_edge().await
         }
 
         /// Wait for the pin to undergo any transition, i.e low to high OR high
         /// to low.
+        ///
+        /// See [Self::wait_for] for more information.
         pub async fn wait_for_any_edge(&mut self) {
             self.pin.wait_for_any_edge().await
         }
     }
 
     #[must_use = "futures do nothing unless you `.await` or poll them"]
-    struct PinFuture {
-        pin_num: u8,
+    struct PinFuture<'d, P: InputPin> {
+        pin: Flex<'d, P>,
     }
 
-    impl PinFuture {
-        fn new(pin_num: u8) -> Self {
-            Self { pin_num }
+    impl<'d, P: InputPin> PinFuture<'d, P> {
+        fn pin_mask(&self) -> u32 {
+            let bank = self.pin.gpio_bank(private::Internal);
+            1 << (self.pin.number() - bank.offset())
+        }
+
+        fn is_done(&self) -> bool {
+            // Only the interrupt handler should clear the async bit, and only if the
+            // specific pin is handling an interrupt.
+            self.pin
+                .gpio_bank(private::Internal)
+                .async_operations()
+                .load(Ordering::Acquire)
+                & self.pin_mask()
+                == 0
         }
     }
 
-    impl core::future::Future for PinFuture {
+    impl<P: InputPin> core::future::Future for PinFuture<'_, P> {
         type Output = ();
 
         fn poll(self: core::pin::Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
-            PIN_WAKERS[self.pin_num as usize].register(cx.waker());
+            PIN_WAKERS[self.pin.number() as usize].register(cx.waker());
 
-            // if pin is no longer listening its been triggered
-            // therefore the future has resolved
-            if !is_listening(self.pin_num) {
+            if self.is_done() {
                 Poll::Ready(())
             } else {
                 Poll::Pending
             }
         }
     }
+
+    impl<P: InputPin> Drop for PinFuture<'_, P> {
+        fn drop(&mut self) {
+            // If the pin isn't listening, the future has either been dropped before setup,
+            // or the interrupt has already been handled.
+            if self.pin.is_listening() {
+                // Make sure the future isn't dropped while the interrupt is being handled.
+                // This prevents tricky drop-and-relisten scenarios.
+
+                set_int_enable(
+                    self.pin.number(),
+                    None, // Do not disable handling pending interrupts.
+                    0,    // Disable generating new events
+                    false,
+                );
+
+                while self.pin.is_interrupt_set() {}
```

Comment:
things that can and do go wrong:

- InterruptExecutor runs at max priority
- InterruptExecutor waits for a pin interrupt
- InterruptExecutor ends up running this code
- code ends up infinite looping, because nothing can handle the pin interrupt